fuel-merkle/src/sparse/proof.rs

```diff
@@ -0,0 +1,795 @@
+use crate::{
+    common::{
+        path::{
+            Instruction,
+            Path,
+        },
+        Bytes32,
+        Prefix,
+        ProofSet,
+    },
+    sparse::{
+        zero_sum,
+        MerkleTreeKey,
+        Node,
+    },
+};
+
+use alloc::vec::Vec;
+use core::{
+    fmt,
+    fmt::Debug,
+};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Proof {
+    Inclusion(InclusionProof),
+    Exclusion(ExclusionProof),
+}
+
+impl Proof {
+    pub fn proof_set(&self) -> &ProofSet {
+        match self {
+            Proof::Inclusion(proof) => &proof.proof_set,
+            Proof::Exclusion(proof) => &proof.proof_set,
+        }
+    }
+
+    pub fn is_inclusion(&self) -> bool {
+        match self {
+            Proof::Inclusion(_) => true,
+            Proof::Exclusion(_) => false,
+        }
+    }
+
+    pub fn is_exclusion(&self) -> bool {
+        !self.is_inclusion()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct InclusionProof {
+    pub proof_set: ProofSet,
+}
+
+impl InclusionProof {
+    pub fn verify(&self, root: &Bytes32, key: &MerkleTreeKey, value: &[u8]) -> bool {
+        let Self { proof_set } = self;
+        let leaf = Node::create_leaf(key.as_ref(), value);
+        let mut current = *leaf.hash();
+        for (i, side_hash) in proof_set.iter().enumerate() {
+            let index = u32::try_from(proof_set.len() - 1 - i).expect("Index is valid");
+            let prefix = Prefix::Node;
+            current = match key.get_instruction(index).expect("Infallible") {
+                Instruction::Left => Node::calculate_hash(&prefix, &current, side_hash),
+                Instruction::Right => Node::calculate_hash(&prefix, side_hash, &current),
+            };
+        }
+        current == *root
+    }
+}
+
+impl Debug for InclusionProof {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let proof_set = self.proof_set.iter().map(hex::encode).collect::<Vec<_>>();
+        f.debug_struct("InclusionProof")
+            .field("Proof set", &proof_set)
+            .finish()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct ExclusionLeaf {
+    leaf_key: Bytes32,
+    leaf_value: Bytes32,
+}
+
+impl ExclusionLeaf {
+    fn leaf_key(&self) -> &Bytes32 {
+        &self.leaf_key
+    }
+
+    fn leaf_value(&self) -> &Bytes32 {
+        &self.leaf_value
+    }
+
+    fn is_placeholder(&self) -> bool {
+        self.leaf_value() == zero_sum()
```

Comment:
We discussed it before, user may use any value, even zero sum. We need to use a separate variant for the case of the placeholder

Reply:
Yes, but in this context, we are looking at the "exclusion leaf" (the nearest included leaf to the excluded key). The `value` of the exclusion leaf is the hash of the user data for this leaf, not the user data itself. This is how this version allows us to use the zero sum for user data.
But, either way, we can use an enum here to simplify this.

Reply:
It is true that, at this point, we have a hash of the user's value. But if the hash is zero, we will treat it as a placeholder, while it is not true. Enum fixes that case.